### ISSUE TITLE
Function Plugs?

### ISSUE THREAD

**Issue by aselder** [akoutmos/unplug#4]
This is amazingly useful...

Is there a way to use function plugs, instead of module plugs?

I'd love to marry Unplug with Plug.BasicAuth.

Normally you specify basic auth using:

`plug :basic_auth, username: "foo", password: "bar"`

but I can figure out how to get this to work with Unplug.

**Comment by akoutmos** [comment]
Unfortunately there is no way to do this at the moment as Unlpug requires the plugs that is calls in if and else to implement the `Plug` behaviour. Fortunately, there is a really easy fix for your use case :).

All you'll need to do is wrap `Plug.BasicAuth.basic_auth(conn, username: username, password: password)` in your own `Plug` module and you are good to go. Something like so:

```elixir
defmodule MyBasicAuthPlug do
  @behaviour Plug

  @impl true
  def init(opts) do
    opts
  end

  @impl true
  def call(conn, _opts) do
    username = System.fetch_env!("AUTH_USERNAME")
    password = System.fetch_env!("AUTH_PASSWORD")
    Plug.BasicAuth.basic_auth(conn, username: username, password: password)
  end
end
```

And then you in your endpoint module you can do something like so:

```elixir
plug Unplug,
    if: {Unplug.Predicates.RequestPathIn, ["/super-secret-path"]},
    do: MyBasicAuthPlug
```

I haven't tested this code and wrote it from memory...but something like that should work :D. Let me know if you can't get that to work!

**Comment by aselder** [comment]
Yeah, that's what I ended up doing, was just hoping there was a nicer way that having to write my own basic auth plug.

Thanks!

**Comment by akoutmos** [comment]
It would be possible if unplug becomes a macro as opposed to a plug. Then it would have the context of the calling module and could `apply(...)` a function against that module. Not something I am currently looking into implementing in Unplug as I would prefer to keep the library free of macros.
